tests/Unit/Architecture/ArchitectureTest.php

```diff
@@ -0,0 +1,143 @@
+                ->that(new ResideInOneOfTheseNamespaces('App\Service\*'))
+                ->should(new NotDependsOnTheseNamespaces('App\Controller\*'))
+                ->because('of the layered architecture'),
+        ];
+
+        self::assertEquals($expectedRules, iterator_to_array($rules));
+    }
+
+    public function test_modular_architecture(): void
+    {
+        $rules = Architecture::withModules()
+            ->module('CRM')->definedBy('App\CRM\*')
+            ->module('InvoiceReconciliation')->definedBy('App\InvoiceReconciliation\*')
+            ->module('Orders')->definedBy('App\Orders\*')
+            ->module('Shared')->definedBy('App\Shared\*')
+            ->module('Bridge')->definedBy('App\Bridge\*')
+
+            ->where('Shared')->shouldNotDependOnAnyModule()
+            ->where('CRM')->mayDependOnModules('Shared')
+            ->where('InvoiceReconciliation')->mayDependOnModules('Shared')
+            ->where('Bridge')->mayDependOnAnyModule()
+            ->where('Orders')->mayDependOnModules('Shared')
+
+            ->rules();
+
+        $expectedRules = [
+            Rule::allClasses()
+                ->that(new ResideInOneOfTheseNamespaces('App\CRM\*'))
+                ->should(new NotDependsOnTheseNamespaces('App\InvoiceReconciliation\*', 'App\Orders\*', 'App\Bridge\*'))
+                ->because('of the modular architecture'),
+            Rule::allClasses()
+                ->that(new ResideInOneOfTheseNamespaces('App\InvoiceReconciliation\*'))
+                ->should(new NotDependsOnTheseNamespaces('App\CRM\*', 'App\Orders\*', 'App\Bridge\*'))
+                ->because('of the modular architecture'),
+            Rule::allClasses()
+                ->that(new ResideInOneOfTheseNamespaces('App\Orders\*'))
+                ->should(new NotDependsOnTheseNamespaces('App\CRM\*', 'App\InvoiceReconciliation\*', 'App\Bridge\*'))
+                ->because('of the modular architecture'),
+            Rule::allClasses()
+                ->that(new ResideInOneOfTheseNamespaces('App\Shared\*'))
+                ->should(new NotDependsOnTheseNamespaces('App\CRM\*', 'App\InvoiceReconciliation\*', 'App\Orders\*', 'App\Bridge\*'))
+                ->because('of the modular architecture'),
+            Rule::allClasses()
+                ->that(new ResideInOneOfTheseNamespaces('App\Bridge\*'))
+                ->should(new NotDependsOnTheseNamespaces())
+                ->because('of the modular architecture'),
+        ];
+
+        self::assertEquals($expectedRules, iterator_to_array($rules));
+    }
+
+    public function test_component_architecture(): void
+    {
+        $rules = Architecture::withComponents()
+            ->component('Doctrine')->definedBy('App\InvoiceReconciliation\Infrastructure\Doctrine\*')
+            ->component('InvoiceStorage')->definedBy('App\InvoiceReconciliation\Infrastructure\InvoiceStorage\*')
+
+            ->where('Doctrine')->shouldNotDependOnAnyComponent()
```

Comment:
My starting idea was to create specific DSLs that solve common use cases in a simpler and more readable way than the more generic syntax we have used so far. 

The syntax proposed by me allows for autocompletion and IMHO it has better readability. On the other hand, it is less consistent with what has been developed so far. 

From the point of view of the simplicity of writing and readability, I would prefer the solution I have implemented. But I don't have a strong opinion on the impact it could have from a coherence point of view. 

I'd like to hear the opinion of @AlessandroMinoccheri and @micheleorselli.

Reply:
( I am okay even with keeping this new style and adapting the old part to this new style, I agree that this new style is great too )